### PR TITLE
Learning rate log callback

### PR DESCRIPTION
**Pull request by SkafteNicki**
# Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure to update the docs?   
- [x] Did you write any new necessary tests?  
- [x] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CHANGELOG.md)?

## What does this PR do?
Fixes https://github.com/PyTorchLightning/pytorch-lightning/issues/1205
Implement simple callback, that logs learning rates during training.

## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/callbacks/lr_logger.py [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+
+Log learning rate for lr schedulers during training
+
+"""
+
+from pytorch_lightning.callbacks.base import Callback
+from pytorch_lightning.utilities.exceptions import MisconfigurationException
+
+
+class LRLogger(Callback):
```
Comment (Borda) [inline pytorch_lightning/callbacks/lr_logger.py:14]: Mind use full name?

File: pytorch_lightning/callbacks/lr_logger.py [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+    Example::
+
+        >>> from pytorch_lightning import Trainer
+        >>> from pytorch_lightning.callbacks import LearningRateLogger
+        >>> lr_logger = LearningRateLogger()
+        >>> trainer = Trainer(callbacks=[lr_logger])
+
+    Logging names are automatically determined based on optimizer class name.
+    In case of multiple optimizers of same type, they will be named `Adam`,
+    `Adam-1` ect. If a optimizer have multiple parameter groups they will
```
Comment (awaelchli) [inline pytorch_lightning/callbacks/lr_logger.py:27]:
```suggestion
    `Adam-1` etc. If a optimizer has multiple parameter groups they will
```

File: pytorch_lightning/callbacks/lr_logger.py [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+    be named `Adam/pg1`, `Adam/pg2` ect. To control naming, pass in a
```
Comment (awaelchli) [inline pytorch_lightning/callbacks/lr_logger.py:28]:
```suggestion
    be named `Adam/pg1`, `Adam/pg2` etc. To control naming, pass in a
```

File: pytorch_lightning/callbacks/lr_logger.py [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+
+            self.names.append(name)
+
+        # Initialize for storing values
+        for name in names:
+            self.lrs[name] = []
+
+    def on_batch_start(self, trainer, pl_module):
+        latest_stat = self._extract_lr(trainer, 'step')
+        if trainer.logger and latest_stat != {}:
```
Comment (awaelchli) [inline pytorch_lightning/callbacks/lr_logger.py:94]:
```suggestion
        if trainer.logger and latest_stat:
```

File: pytorch_lightning/callbacks/lr_logger.py [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+            trainer.logger.log_metrics(latest_stat, step=trainer.global_step)
+
+    def on_epoch_start(self, trainer, pl_module):
+        latest_stat = self._extract_lr(trainer, 'epoch')
+        if trainer.logger and latest_stat != {}:
```
Comment (awaelchli) [inline pytorch_lightning/callbacks/lr_logger.py:99]:
```suggestion
        if trainer.logger and latest_stat:
```

File: pytorch_lightning/callbacks/lr_logger.py [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+            trainer.logger.log_metrics(latest_stat, step=trainer.global_step)
+
+    def _extract_lr(self, trainer, interval):
+        """ Extracts learning rates for lr schedulers and save information
```
Comment (awaelchli) [inline pytorch_lightning/callbacks/lr_logger.py:103]:
```suggestion
        """ Extracts learning rates for lr schedulers and saves information
```

File: pytorch_lightning/callbacks/lr_logger.py [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+            into dict structure. """
+        latest_stat = {}
+        for name, scheduler in zip(self.names, trainer.lr_schedulers):
```
Comment (awaelchli) [inline pytorch_lightning/callbacks/lr_logger.py:106]: May I suggest a test that the names are correctly assigned?

File: CHANGELOG.md [hunk 8]
```diff
@@ -9,9 +9,9 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 ### Added
 
 - Added `auto_select_gpus` flag to trainer that enables automatic selection of available GPUs on exclusive mode systems.
-- Added learning rate finder ([#1347](https://github.com/PyTorchLightning/pytorch-lightning/pull/1347))
-
```
Comment (awaelchli) [inline CHANGELOG.md:13]: I think you need to keep the blank lines between items to prevent merging issues

File: pytorch_lightning/callbacks/lr_logger.py [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+                opt_name = 'lr-' + sch.optimizer.__class__.__name__
+                name = opt_name
+                counter = 0
+                # Multiple schduler of the same type
+                while True:
+                    counter += 1
+                    if name in names:
+                        name = opt_name + '-' + str(counter)
+                    else:
+                        break
```
Comment (Borda) [inline pytorch_lightning/callbacks/lr_logger.py:75]:
```suggestion
                    if name not in names:
                        break
                    name = opt_name + '-' + str(counter)
```

File: pytorch_lightning/callbacks/lr_logger.py [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+            into dict structure. """
+        latest_stat = {}
+        for name, scheduler in zip(self.names, trainer.lr_schedulers):
+            if scheduler['interval'] == interval:
+                param_groups = scheduler['scheduler'].optimizer.param_groups
+                if len(param_groups) != 1:
+                    for i, pg in enumerate(param_groups):
+                        lr = pg['lr']
+                        self.lrs[name + '/' + str(i + 1)].append(lr)
+                        latest_stat[name + '/' + str(i + 1)] = lr
```
Comment (Borda) [inline pytorch_lightning/callbacks/lr_logger.py:113]:
```suggestion
                        key = f'{name}/{i + 1}'
                        self.lrs[key].append(lr)
                        latest_stat[key] = lr
```

File: tests/trainer/test_callbacks.py [hunk 11]
```diff
@@ -181,3 +182,59 @@ def training_step(self, *args, **kwargs):
...
+        max_epochs=5,
+        val_percent_check=0.1,
+        train_percent_check=0.5,
+        callbacks=[lr_logger]
+    )
+
+    trainer = Trainer(**trainer_options)
+    results = trainer.fit(model)
+
+    assert lr_logger.lrs != {}, 'No learning rates logged'
```
Comment (Borda) [inline tests/trainer/test_callbacks.py:209]:
```suggestion
    assert lr_logger.lrs, 'No learning rates logged'
```

File: tests/trainer/test_callbacks.py [hunk 12]
```diff
@@ -181,3 +182,59 @@ def training_step(self, *args, **kwargs):
...
+    assert len(lr_logger.lrs.keys()) == len(trainer.lr_schedulers), \
```
Comment (Borda) [inline tests/trainer/test_callbacks.py:210]:
```suggestion
    assert len(lr_logger.lrs) == len(trainer.lr_schedulers), \
```

File: tests/trainer/test_callbacks.py [hunk 13]
```diff
@@ -181,3 +182,59 @@ def training_step(self, *args, **kwargs):
...
+    lr_logger = LearningRateLogger()
+    trainer_options = dict(
+        default_root_dir=tmpdir,
+        max_epochs=1,
+        val_percent_check=0.1,
+        train_percent_check=0.5,
+        callbacks=[lr_logger]
+    )
+
+    trainer = Trainer(**trainer_options)
```
Comment (Borda) [inline tests/trainer/test_callbacks.py:235]:
```suggestion
    trainer = Trainer(
        default_root_dir=tmpdir,
        max_epochs=1,
        val_percent_check=0.1,
        train_percent_check=0.5,
        callbacks=[lr_logger]
    )
```

File: tests/trainer/test_callbacks.py [hunk 14]
```diff
@@ -181,3 +182,59 @@ def training_step(self, *args, **kwargs):
...
+    lr_logger = LearningRateLogger()
+    trainer_options = dict(
+        default_root_dir=tmpdir,
+        max_epochs=5,
+        val_percent_check=0.1,
+        train_percent_check=0.5,
+        callbacks=[lr_logger]
+    )
+
+    trainer = Trainer(**trainer_options)
```
Comment (Borda) [inline tests/trainer/test_callbacks.py:206]:
```suggestion
    trainer = Trainer(
        default_root_dir=tmpdir,
        max_epochs=5,
        val_percent_check=0.1,
        train_percent_check=0.5,
        callbacks=[lr_logger]
    )
```

File: tests/trainer/test_callbacks.py [hunk 15]
```diff
@@ -181,3 +182,59 @@ def training_step(self, *args, **kwargs):
...
+    results = trainer.fit(model)
+
+    assert lr_logger.lrs != {}, 'No learning rates logged'
+    assert len(lr_logger.lrs.keys()) == len(trainer.lr_schedulers), \
```
Comment (Borda) [inline tests/trainer/test_callbacks.py:239]:
```suggestion
    assert lr_logger.lrs, 'No learning rates logged'
    assert len(lr_logger.lrs) == len(trainer.lr_schedulers), \
```

File: pytorch_lightning/callbacks/lr_logger.py [hunk 16]
```diff
@@ -0,0 +1,117 @@
...
+            else:
+                opt_name = 'lr-' + sch.optimizer.__class__.__name__
+                name = opt_name
+                counter = 0
+                # Multiple schduler of the same type
+                while True:
+                    counter += 1
+                    if name not in names:
+                        break
+                    name = opt_name + '-' + str(counter)
```
Comment (Borda) [inline pytorch_lightning/callbacks/lr_logger.py:74]:
```suggestion
                i, name = 1, opt_name
                # Multiple schduler of the same type
                while True:
                    if name not in names:
                        break
                    i, name = i + 1, f'{opt_name}-{i}'
```

File: pytorch_lightning/callbacks/lr_logger.py [hunk 17]
```diff
@@ -0,0 +1,115 @@
...
+        """ Extracts learning rates for lr schedulers and saves information
+            into dict structure. """
+        latest_stat = {}
+        for name, scheduler in zip(self.names, trainer.lr_schedulers):
+            if scheduler['interval'] == interval:
+                param_groups = scheduler['scheduler'].optimizer.param_groups
+                if len(param_groups) != 1:
+                    for i, pg in enumerate(param_groups):
+                        lr = pg['lr']
+                        key = f'{name}/{i + 1}'
```
Comment (Borda) [inline pytorch_lightning/callbacks/lr_logger.py:109]:
```suggestion
                        lr, key = pg['lr'], f'{name}/{i + 1}'
```

File: pytorch_lightning/callbacks/lr_logger.py [hunk 18]
```diff
@@ -0,0 +1,114 @@
...
+            # Multiple param groups for the same schduler
+            param_groups = sch.optimizer.param_groups
+            if len(param_groups) != 1:
+                for i, pg in enumerate(param_groups):
+                    temp = name + '/pg' + str(i + 1)
+                    names.append(temp)
+            else:
+                names.append(name)
+
+            self.names.append(name)
```
Comment (awaelchli) [inline pytorch_lightning/callbacks/lr_logger.py:83]: maybe this could be refactored into a function that takes the scheduler as input and outputs the names.

File: pytorch_lightning/callbacks/lr_logger.py [hunk 19]
```diff
@@ -0,0 +1,114 @@
...
+
+        # Initialize for storing values
+        for name in names:
+            self.lrs[name] = []
```
Comment (awaelchli) [inline pytorch_lightning/callbacks/lr_logger.py:87]: perhaps` self.lrs` could be a `defaultdict`? Then this for loop would not be needed.

**Comment by ivannz** [comment]
Thank you for this extremely useful callback, @SkafteNicki  !

I've noticed that you are using `d = dict.fromkeys` in [lr_logger.py#L62](https://github.com/PyTorchLightning/pytorch-lightning/blob/caa9c6760b282637c5f2e7408f85685cc3b342cf/pytorch_lightning/callbacks/lr_logger.py#L62) and later use `d[key].append(...)` on [lr_logger.py#L84-L87](https://github.com/PyTorchLightning/pytorch-lightning/blob/caa9c6760b282637c5f2e7408f85685cc3b342cf/pytorch_lightning/callbacks/lr_logger.py#L84-L87). I think `L62` is incorrect, since `L84-L87` would appends to **all lists** at once, because the dictionary created at line `L62` contains reference to the **same** list. The following example demonstrates the issue:
```python
names = ['a', 'b', 'c']

d1 = dict.fromkeys(names, [])
d2 = {k: [] for k in names}

d1['a'].append(1)
d1['c'].append(2)

d2['a'].append(1)
d2['c'].append(2)

print(f'd1 = {d1}')
print(f'd2 = {d2}')
```
prints
```
d1 = {'a': [1, 2], 'b': [1, 2], 'c': [1, 2]}
d2 = {'a': [1], 'b': [], 'c': [2]}
```
Contents of `d1` do not seem to be the behaviour desired by the logic of [._extract_lr()](https://github.com/PyTorchLightning/pytorch-lightning/blob/caa9c6760b282637c5f2e7408f85685cc3b342cf/pytorch_lightning/callbacks/lr_logger.py#L74).

**Comment by SkafteNicki** [comment]
@ivannz good catch, I was not aware that `fromkeys` should not be used in combination with mutable objects. Could you do a PR to correct this?

**Comment by ivannz** [comment]
@SkafteNicki Thank you, done.

As a sidenote: in my experiments I found it quite useful to disable the [misconfiguration check](https://github.com/PyTorchLightning/pytorch-lightning/blob/8ca8336ce52ee7379f4d399520636143eb31018b/pytorch_lightning/callbacks/lr_logger.py#L48-L52), since it requires an inconvenient branching (to remove  `LearningRateLogger`) before a call to `trainer` when no scheduler is present.

What was the rationale behind issuing an `MisconfigurationException` rather than a warning, that would still allow for the trainer to continue?

**Comment by SkafteNicki** [comment]
@ivannz not really any rational behind, I just assumed that everyone that used this has a scheduler in their model. However, I can clearly see that it is smarter to have a warning instead of an exception. Could you also add this to the PR?

**Comment by ivannz** [comment]
@SkafteNicki done. I am going to patch both in a single PR.

**Comment by yitang** [comment]
how to access the LR logged by LearningRateLogger? i couldn't find an example from the docs.

**Comment by SkafteNicki** [comment]
@yitang the learning rate(s) are stored in the the `lrs` attribute of the `LearningRateLogger`

**Comment by williamFalcon** [comment]
@SkafteNicki is this written in the docs?

**Comment by SkafteNicki** [comment]
@williamFalcon No, as the (originally) intention of the `LearningRateLogger` is that the lrs are logged as any other scalar to the `trainer.logger`, thus the user could interact with the values in that way. However, maybe it could be made clear in the docs that they are also stored in the `lrs` attribute.
